binstat/README.md

```diff
@@ -0,0 +1,175 @@
+# `binstat` - run-time & space efficient stats in bins
+
+## Why use `binstat`?
+
+* `pprof` reports the seconds of CPU used and not the seconds of wall-clock used.
+* `trace` can be too heavyweight for larger Golang programs.
+* prometheus can be heavyweight, and it's difficult to programmatically manage stats once exported.
+* logging can be heavyweight; 1M function calls == 1M log lines.
+
+## `binstat` goals
+
+* Be able to instrument arbitrary code points without worrying about reducing run-time performance.
+* Run instrumented code points arbitrary times without worrying about reducing run-time performance.
+* Allow collected stats to be managed and compared programmatically; is this worse/better than last run?
+
+## How `binstat` works?
+
+* Sprinkle `binstat.New*(<what>)` and `binstat.End*()` into Golang source code, e.g.:
+
+```
+	for loop := 0; loop < 2; loop++ {
+		p := binstat.NewTime(fmt.Sprintf("loop-%d", loop), "")
+		myFunc()
+		binstat.End(p)
+	}
+```
+
+* Run the program, and max. once per second a stats file will be written opportunistically, e.g.:
+
+```
+$ cat myproc.pid-085882.txt| egrep loop
+/GOMAXPROCS=1,GC=100/what[loop-0]/file[mypkg.myFunc:161]/time[1.000000-1.999999]=1 1.912292
+/GOMAXPROCS=8,GC=100/what[loop-1]/file[mypkg.myFunc:161]/time[0.600000-0.699999]=1 0.696432
```

Comment:
binstat varies `gomaxprocs` by itself?

Reply:
Yes, one of the tests does something similar / inspired by this pprof++ test Golang script here [1], i.e. n Go-routines run in parallel, each takes the same CPU to run, they are run several times and with different `GOMAXPROCS` and with regular pprof profiling. The outputs for each run are summarized.

Note: `binstat` itself does not change `GOMAXPROCS` at run-time but does report its current value in case it changes since this can affect wall-clock performance dramatically...

[1] https://eng.uber.com/pprof-go-profiler/